ggu/main.c

```diff
@@ -16,11 +16,295 @@
+			sizeof(policy->flow.f.v6));
+	else {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: parse policy with unknown network layer protocol %hhu\n",
+			policy->flow.proto);
+		return -1;
+	}
+
+	rte_memcpy(&policy->params, policy_ptr, sizeof(policy->params));
+
+	return 0;
+}
+
+static void
+process_single_policy(struct ggu_policy *policy, struct ggu_config *ggu_conf)
+{
+	/*
+	 * Identify which GK block is responsible for the
+	 * pair <Src, Dst> in the decision.
+	 */
+	int block_id = get_responsible_block(&policy->flow, ggu_conf->gk);
+	struct gk_cmd_entry entry;
+
+	if (block_id < 0)
+		return;
+	/*
+	 * Obtain mailbox of that GK block,
+	 * and send the policy decision to the GK block.
+	 */
+	entry.op = GGU_POLICY_ADD;
+	rte_memcpy(&entry.u.ggu.flow, &policy->flow, sizeof(policy->flow));
+	entry.u.ggu.params.state = policy->params.state;
+	entry.u.ggu.params.tx_rate_kb_sec =
+		rte_be_to_cpu_32(policy->params.tx_rate_kb_sec);
+	entry.u.ggu.params.cap_expire_sec =
+		rte_be_to_cpu_32(policy->params.cap_expire_sec);
+	entry.u.ggu.params.renewal_step_ms =
+		rte_be_to_cpu_32(policy->params.renewal_step_ms);
+
+	send_gk_cmd(block_id, &entry, ggu_conf->gk);
+}
+
+static void
+process_single_packet(struct rte_mbuf *pkt, struct ggu_config *ggu_conf)
+{
+	uint8_t j;
+	uint8_t *data;
+	uint8_t *policy_ptr;
+	uint8_t num_policies;
+	uint16_t *proto_ptr;
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr *ip4hdr;
+	struct ipv6_hdr *ip6hdr;
+	struct udp_hdr *udphdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+		ip4hdr = rte_pktmbuf_mtod_offset(pkt, 
+			struct ipv4_hdr *, sizeof(struct ether_hdr));
+		if (ip4hdr->next_proto_id != IPPROTO_UDP) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received non-UDP packets, ntuple filter bug!\n");
+			goto free_packet;
+		}
+
+		udphdr = (struct udp_hdr *)&ip4hdr[1];
+		break;
+
+	case ETHER_TYPE_IPv6:
+		ip6hdr = rte_pktmbuf_mtod_offset(pkt, 
+			struct ipv6_hdr *, sizeof(struct ether_hdr));
+		if (ip6hdr->proto != IPPROTO_UDP) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received non-UDP packets, ntuple filter bug!\n");
+			goto free_packet;
+		}
+
+		udphdr = (struct udp_hdr *)&ip6hdr[1];
+		break;
+
+	default:
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: unknown network layer protocol %hu\n",
+			ether_type);
+		goto free_packet;
+		break;
+	}
+
+	if (udphdr->src_port != rte_cpu_to_be_16(ggu_conf->ggu_src_port) ||
+			udphdr->dst_port !=
+			rte_cpu_to_be_16(ggu_conf->ggu_dst_port)) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: unknown udp src port %hu, dst port %hu, ntuple filter bug!\n",
+			rte_be_to_cpu_16(udphdr->src_port),
+			rte_be_to_cpu_16(udphdr->dst_port));
+		goto free_packet;
+	}
+
+	data = (uint8_t *)&udphdr[1];
+	num_policies = data[0];
+	data = data + 1;
+	proto_ptr = (uint16_t *)data;
+	policy_ptr = (uint8_t *)((uint16_t *)data + num_policies);
```

Comment:
Use the size of the payload available in the UDP header to check that the expected size and the payload matches. Otherwise, malformed truncated packets can lead the code to misbehave.

Adding a XXX comment to check UDP checksum is also a good idea.